scalems/wrappers/__init__.py

```diff
@@ -16,32 +15,109 @@
         arguments to a *modify_input* command.
 
+        For most options, then a keyword argument can directly
+        parallel the keyword syntax used by the simulation programs
+        themselves: LAMMPS, PLUMED, GROMACS (in the .mdp) all use
+        keyword options.  It's not as clear for OpenMM, but we should
+        use the ExTASY implementations to guide us as to how modification is done there.  
+
+        One question is how the defintion of the chemical system
+        should changed using this approach. This is not set up very
+        well in the keyword format.  We could, for now, just replace
+        the file containing the molecular system definition entirely
+        using a modify_input.  This might be the best bet for now, but
+        we will want to think about what sorts of changes might be of
+        interest, if any in the future.
+
+        The issue is that in most cases, if the chemical system is
+        changing, it usually requires changes of many lines of
+        topologies, and so passing it in as a dictionary will likely
+        be a big pain. Likely easier to take care of it outside the
+        simulation for now. Eventually, if some parallel work on
+        simulation topology definitions in Open Force Field works,
+        then it may be possible to do this better.
+
+        Things like "training potential" that get attached in the
+        examples should probably be attached at this point in the
+        workflow, rather than at the simulate. For most workflows, it
+        could ONLY be added at this stage, since potentials are
+        defined in the input files.  For GMX_API driven simulations,
+        the actual attachment of the potentials to the md simulation
+        could happen in the simulate function, but it would be better
+        if they were loaded into the data structure in the
+        modify_input sage.
+
     simulate(simulation_input: SimulationInput, **kwargs) -> Simulation:
         Create a simulation operation instance. Accepts input from *make_input*
         or *modify_input*, along with key word arguments specified by the wrapper
         module.
 
+        Potentially, we use this to specify which resources assoicated
+        with the job?  The goal of this function is to define a task
+        that is ready to dispatch for execution.  
+
+        MRS: some notes that I took during the conversation last week:
+        Any of the things that affect the checkpointable identity are
+        immutable after simulate.  So after simulate, define an
+        equivalence class. Operate by fingerprinting commands and data
+        to determine identity that should distinguish it.  Workflow
+        level checkpointing.
+       
+        MRS: I'm still a little confused by the point of this.  I
+        assume the importance of checkpointing is to know whether one
+        should relaunch a simulation if it fails?  And if it's not
```

Comment:
It seems like having explicit metadata for the tasks would be the best way to do it? A simulate task can have associated metadata with it, and it would then be a matter of figuring out what functions were in charge of marking that metadata as completed (or some other status).  But it seems like trying to figure out what to do with a simulation task by direct inspection of all of the data in it would be really hard.  Seems like it would be simpler to have the task handling trust what the metadata said about that task if we can figure out a scheme for the right time/responsibilities for labeling.

Reply:
Yes, the questions I am asking are in support of identifying the structure and meaning of the metadata.

There is an interesting long-term question of whether we want to allow or encourage researchers to express their add-on codes abstractly, to be identified independently from the implementation. But what we have assumed so far is to assume that a package name like `jmhays.brer_plugin` is pretty much all the information we have about the method we are attaching unless the researcher wants to provide their own API for labeling the logic it contains.